<h2>Exploring Ebay Car Sales Data</h2>

We will be working on a dataset of used cars from eBay Kleinanzeigen, the classified ads section of the German eBay website.

The dataset was originally scraped and uploaded to :<br>

<a href="https://www.kaggle.com/orgesleka/used-cars-database/data">Kaggle</a>

This version of the dataset is a sample of 50,000 data points and was prepared by Dataquest to be less clean than the Kaggle version of the data, in order to more closely resemble what you would expect from a scraped dataset.


The dataset's metadata/data dictionary:

&nbsp; | &nbsp;
------ | -----
dateCrawled  | When this ad was first crawled. All field-values are taken from this date.
When this ad was first crawled | All field-values are taken from this date.
    name | Name of the car.
    seller | Whether the seller is private or a dealer.
    offerType | The type of listing
    price | The price on the ad to sell the car.
    abtest | Whether the listing is included in an A/B test.
    vehicleType | The vehicle Type.
    yearOfRegistration | The year in which which year the car was first registered.
    gearbox | The transmission type.
    powerPS | The power of the car in PS.
    model | The car model name.
    kilometer | How many kilometers the car has driven.
    monthOfRegistration | The month in which which year the car was first registered.
    fuelType | What type of fuel the car uses.
    brand | The brand of the car.
    notRepairedDamage | If the car has a damage which is not yet repaired.
    dateCreated | The date on which the eBay listing was created.
    nrOfPictures | The number of pictures in the ad.
    postalCode | The postal code for the location of the vehicle.
    lastSeenOnline | When the crawler saw this ad last online.


This is mainly a data-cleaning project, but also uses pandas to analyse the used car listings. We start by importing the libraries we need and reading the dataset from the CSV file into pandas.

In [209]:
""" When attempting to read the CSV file into a pandas dataframe(df) 
    without specifying any encoding (which defaults to UTF-8)
    a UnicodeDecodeError error was returned and so 
    the next two most popular encodings - Latin-1 and Windows-1252
    were attempted with: Latin-1 not returning error """


#import libraries
import pandas as pd
import numpy as np

#read file into pandas and assign to 'autos' variable
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [210]:
#render the first few and last few values of the pandas object
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [211]:
#DataFrame.info() method is used to get a summary of column count and datatype
#DataFrame.head() method is used to print information about the autos dataframe

autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50


<h2> Observations </h2>

The dataset has 20 columns, most of which are the object datatype aka dtype which are stored as strings. 
Since the column names use camelcase instead of Python's preferred snakecase we can't just replace spaces with underscores, so we'll start by cleaning the column names to make the data easier to work with and and reword some of the column names based on the data dictionary to be more descriptive. 

To start, we'll use the df.columns attribute to print an array of the existing column names, then copy that array and make certain edits to columns names.

 <ul>
  <li>yearOfRegistration to registration_year</li>
  <li>notRepairedDamage to unrepaired_damage</li>
  <li>dateCreated to ad_created</li>
  <li>The rest of the columnn names from camelcase to snakecase.</li>
</ul> 

    
Then the edited column names will be assigned back to the df.columns attribute.

<h2> Rename columns</h2>

In [212]:
# assign list of new column names to the df.
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

# check new columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50


<h2>Initial Data Exploration and Cleaning</h2>

Now that the column names have been cleaned we'll do some more exploration to determine what other cleaning tasks need to be done.
Initially, we'll look for:

 <ul>
  <li>    Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.</li>
  <li>    Examples of numeric data stored as text which can be cleaned and converted.</li>
</ul> 

We'll use the following methods for exploring the data: - DataFrame.describe() (with include='all' to get a summary of both categorical and numeric columns as the default is only to summarise the numerical columns) - Series.value_counts() and Series.head() if any columns need a closer look.


In [213]:
# check descriptive statistics for all columns.
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-11 22:38:16  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN



We now see that:
<ol>
  <li>Of the 20 columns there are a number of text columns where almost all of the values are the same:
    <ol>
      <li>seller</li>
      <li>offer_type</li>
    </ol>
  </li>
  <li>The num_photos column has min and max of zero, we'll need to investigate this further.</li>
   <li>The registration_year data has min and max years of 1000 and 9999 which seems incorrect and maybe the result of typos.</li>
</ol>

In [214]:
# .value_counts() to return a Series containing counts of unique values
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

The num_photos column has only zeroes so we'll drop this column, along with 
    seller &
    offer_type
as the values are mostly the same and so they won't have useful information for this analysis.

In [215]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)


The price and odometer columns are numeric with extra characters being stored as text. So we need to convert their dtype and rid them of extraneous characters in order to do numeric operations. We shall :

 <ul>
  <li> Clean any non-numeric characters -- '$' and ',' .</li>
  <li> Convert the df column to the integer dtype, by calling the astype() method which enables explicit conversion.</li>
    <li> Call df.rename() on the df to rename the column to odometer_km. </li>
</ul> 

In [216]:

# Removes $ and comma and converts to int dtype
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [217]:

# renames odometer column and removes characters -- commas and km
autos["odometer"] = (autos["odometer"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

<h2>Exploring Odometer Readings and Price Columns</h2>

Above, we learned that there are a number of text columns where almost all of the values are the same (seller and offer_type). We also converted the price and odometer columns to numeric types and renamed odometer to odometer_km.

We'll now  look for data that doesn't look right in the odometer_km and price columns.

We'll check the columns' minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.
We'll use:

<ul>
    <li>Series.unique().shape to see how many unique values</li>
    <li>Series.describe() to view descriptive statistics - min/max/median etc</li>
    <li>Series.value_counts(), to get a Series containing counts of unique values, with some variations:
        <ul>
            <li>chained to .head() if there are lots of values.</li>
            <li>Because Series.value_counts() returns a series, we can use Series.sort_index() with ascending= True or False to view the highest and lowest values with their counts (can also chain to head() here).</li>
        </ul>
    </li>
</ul>

In [218]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

From this we observe that the values in this field are rounded and so it might be reasoned that sellers had to choose from pre-set options for this field. Also there are more high mileage than low mileage vehicles. Although mileage alone can be misleading in the valuation.  In Germany cars do an 8300-miles / year average.

In [219]:
#shape -returns a tuple of the DataFrame's dimensions. unique()to see how many unique values
print(autos["price"].unique().shape)

print(autos["price"].describe())

# value_counts() sorts and displays the frequency in descending order 
# so that the first element is the most frequently-occurring element.head(20) - first 20 rows
autos["price"].value_counts().head(20)



(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

Again, the prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.

Highest frequency (1421) is a price of zero -  given that this is only 2% of the of the cars, we might consider removing these rows. Let's further check the highest prices.

In [220]:
# sort_index --- sorts object by labels along --- rows - when ascending=False

autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [221]:

# sort_index --- sorts object by labels along columns -- ascending=True is default argument.

autos["price"].value_counts().sort_index().head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

<h2>Outliers</h2>

For the price column:

The prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.

Given that eBay is an auction site, there could legitimately be items where the opening bid is one dollar. So we'll keep the one dollar items, but remove those priced above 350K since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

There are 1,421 cars listed with $0 price, given that this is only two percent of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot.

Let's further look at the frequency counts of the highest prices in the range 1 - 350,000 and the centiles they fall into to get a better sense of the data's distribution.

In [222]:
# check outliers
# df[df["col"].between(x,y)] -- returns a boolean Series equivalent to left <= series <= right.

autos = autos[autos["price"].between(1,350000)]
autos["price"].describe().apply(np.ceil)

count     48565.0
mean       5889.0
std        9060.0
min           1.0
25%        1200.0
50%        3000.0
75%        7490.0
max      350000.0
Name: price, dtype: float64


<h2>Exploring the date columns</h2>

Let's now move on to the date columns and understand the date range the data covers.

There are five columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

<ol>
<li> date_crawled: added by the crawler</li>
<li> last_seen: added by the crawler</li>
<li> ad_created: from the website</li>
    <br>
<li> registration_month: from the website</li>
<li> registration_year: from the website</li>
    </ol>

1, 2 & 3 are all string values right now, so we need to convert them into a numerical dtype in order to do analysis. The other two are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without any extra data processing.

We'll explore each of these columns to learn more about the listings. Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [223]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The first 10 characters in each column represent the day. For all the columns above, let’s calculate the density distribution as percentages instead of counts.

In [228]:
"""To understand the date range, we extract just the date values, 
   using Series.value_counts() to generate a distribution
   and then sort by the index. It excludes NA values by default so this is set to False...
   The first 10 characters in the column are selected -- Series.str[:10]:"""

(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [229]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.


In [230]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64



The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.


In [231]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months. Now we'll use Series.describe() to understand the distribution of the registration_year, since it contains some outliers.

In [233]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


<h2>Dealing with Incorrect Registration Year Data</h2>

The minimum value is 1000, before cars were invented, maximum being 9999 - many years into the future. Because a car can't first be registered before the listing is seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely:


In [234]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

Given that this is less than 4% of our data, we will remove these rows.

In [236]:

"""We'll select select rows that fall within the above value range
   and remove those outside those upper and lower bounds 
   and calculate the distribution of 
   the remaining values using Series.value_counts(normalize=True)."""

autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)



2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

<h2>Exploring Price by Brand</h2>



In [238]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r



German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

As Series.value_counts() produces a series with index labels, we'll use the Series.index attribute to access the labels.

In [239]:
brand_counts = autos["brand"].value_counts(normalize = True)
common_brands = brand_counts[brand_counts > 0.05].index
print(common_brands)


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [240]:
""" An empty dictionary is used to hold the aggregate data.
    Then selected brands are iterated with the mean price assigned to the dictionary,
    with the brand name as the key"""

brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}



Having aggregated across brands to understand the mean price, we see that of the top 6 brands, there is a distinct price gap:

<ul>
    <li>Audi, BMW and Mercedes Benz are more expensive</li>
    <li>Ford and Opel are less expensive</li>
    <li>Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.</li>
    </ul>



<h2>Exploring Mileage</h2>

For these top 6 brands, we'll use aggregation to understand the average mileage and if there's any link with the mean price. 

To do this, we'll combine data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To implement in code we'll use the panda.series constructor to return a one-dimensional ndarray with axis labels and the pandas.dataframe constructor.

In [247]:
bmp_series = pd.Series(brand_mean_prices)

print(bmp_series)

volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
dtype: int64


The keys in the dictionary have now become the index in the series object. We can now create a single-column dataframe from this series object. We need to use the columns parameter when calling the dataframe constructor (which takes an array-like object as an argument - dict, list, ndarray, series etc) to specify the column name (or the column name will be set to 0 by default):

In [248]:
# pandas.DataFrame( data, index, columns, dtype, copy)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [251]:
"""Calculates the 
   mean mileage and mean price for each of the top brands, 
   storing the results in a dictionary."""


brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

#Convert both dictionaries to series objects, using the series constructor.
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [252]:
#Create a dataframe from the first series object using the dataframe constructor.

brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [253]:
#Assign the other series as a new column in this dataframe.

brand_info["mean_price"] = mean_prices
brand_info



mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

<h3>Conclusion</h3>

The range of car mileages do not vary as much as the prices do by brand. Rather, all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.